pkg/plugins/policies/matchers/match.go

```diff
@@ -13,44 +13,101 @@ import (
 	xds_context "github.com/kumahq/kuma/pkg/xds/context"
 )
 
-type ResourceSpecWithTargetRef interface {
-	core_model.ResourceSpec
-	GetTargetRef() *common_proto.TargetRef
-}
-
 // MatchedPolicies match policies using the standard matchers using targetRef (madr-005)
 func MatchedPolicies(rType core_model.ResourceType, dpp *core_mesh.DataplaneResource, resources xds_context.Resources) (core_xds.TypedMatchingPolicies, error) {
 	policies := resources.ListOrEmpty(rType)
 
-	matchedPolicies := []core_model.Resource{}
+	matchedPoliciesByInbound := map[core_xds.InboundListener][]core_model.Resource{}
+	dpPolicies := []core_model.Resource{}
+
 	for _, policy := range policies.GetItems() {
-		spec, ok := policy.GetSpec().(ResourceSpecWithTargetRef)
+		spec, ok := policy.GetSpec().(core_xds.ResourceSpecWithTargetRef)
 		if !ok {
 			return core_xds.TypedMatchingPolicies{}, errors.Errorf("resource type %v doesn't support TargetRef matching", rType)
 		}
 
 		targetRef := spec.GetTargetRef()
-		if isDataplaneSelectedByTargetRef(targetRef, dpp) {
-			matchedPolicies = append(matchedPolicies, policy)
+		selectedInbounds := inboundsSelectedByTargetRef(targetRef, dpp)
+		if len(selectedInbounds) == 0 {
+			// DPP is not matched by the policy
+			continue
+		}
+
+		dpPolicies = append(dpPolicies, policy)
+
+		for _, inbound := range selectedInbounds {
+			matchedPoliciesByInbound[inbound] = append(matchedPoliciesByInbound[inbound], policy)
 		}
 	}
 
-	sort.Sort(ByTargetRef(matchedPolicies))
+	sort.Sort(ByTargetRef(dpPolicies))
+
+	for _, ps := range matchedPoliciesByInbound {
+		sort.Sort(ByTargetRef(ps))
+	}
 
 	return core_xds.TypedMatchingPolicies{
 		Type:              rType,
-		DataplanePolicies: matchedPolicies,
+		DataplanePolicies: dpPolicies,
+		FromRules: core_xds.FromRules{
+			Rules: fromRules(matchedPoliciesByInbound),
+		},
+		ToRules: core_xds.ToRules{
+			Rules: toRules(dpPolicies),
+		},
 	}, nil
 }
 
-func isDataplaneSelectedByTargetRef(tr *common_proto.TargetRef, dpp *core_mesh.DataplaneResource) bool {
+func fromRules(
+	matchedPoliciesByInbound map[core_xds.InboundListener][]core_model.Resource,
+) map[core_xds.InboundListener]core_xds.Rules {
+	rules := map[core_xds.InboundListener]core_xds.Rules{}
+	for inbound, policies := range matchedPoliciesByInbound {
+		fromList := []core_xds.PolicyItem{}
+		for _, p := range policies {
+			policyWithFrom, ok := p.GetSpec().(interface{ GetFromList() []core_xds.PolicyItem })
```

Comment:
I think this should be explicit interface defined next to `ResourceSpecWithTargetRef` so there is a set of interfaces that policy needs to implement in order to have specific matching.

WDYT?

Reply:
Created 2 interfaces:
* PolicyWithToList
* PolicyWithFromList